vmware-base/src/com/cloud/hypervisor/vmware/util/VmwareClient.java

```diff
@@ -353,10 +356,36 @@ public boolean waitForTask(ManagedObjectReference task) throws InvalidPropertyFa
             if (result[1] instanceof LocalizedMethodFault) {
                 throw new RuntimeException(((LocalizedMethodFault) result[1]).getLocalizedMessage());
             }
-        } catch(WebServiceException we) {
-            s_logger.debug("Cancelling vCenter task because task failed with " + we.getLocalizedMessage());
+        } catch (WebServiceException we) {
+            s_logger.warn("Session to vCenter failed with: " + we.getLocalizedMessage());
+
+            TaskInfo taskInfo = (TaskInfo)getDynamicProperty(task, "info");
+            if (!taskInfo.isCancelable()) {
+                s_logger.warn("vCenter task: " + taskInfo.getName() + "(" + taskInfo.getKey() + ")" + " will continue to run on vCenter because the task cannot be cancelled");
+                throw new RuntimeException(we.getLocalizedMessage());
+            }
+
+            s_logger.debug("Cancelling vCenter task: " + taskInfo.getName() + "(" + taskInfo.getKey() + ")");
             getService().cancelTask(task);
-            throw new RuntimeException("vCenter task failed due to " + we.getLocalizedMessage());
+
+            // Since task cancellation is asynchronous, wait for the task to be cancelled
+            Object[] result = waitForValues(task, new String[] {"info.state", "info.error"}, new String[] {"state"},
```

Comment:
@sureshanaparti I see that however my comment aimed to encourage defensive programming. Maybe now the code works, but it can still improve upon overall elegance and include additional checks to defend future bugs [1].

[1] https://en.wikipedia.org/wiki/Defensive_programming